src/main/scala/inox/utils/FutureUtils.scala

```diff
@@ -0,0 +1,30 @@
+/* Copyright 2009-2021 EPFL, Lausanne */
+
+package inox.utils
+
+import scala.util._
+import scala.concurrent._
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext.Implicits.global
+import java.util.concurrent.atomic.AtomicInteger
+
+object FutureUtils {
+
+  def findFirst[T](futures: Seq[Future[T]])(cond: T => Boolean): Option[T] = {
+    val p = Promise[Option[T]]()
+    val n = futures.length
+    val i = new AtomicInteger(0)
+
+    for (future <- futures) {
+      future.onComplete((result: Try[T]) => result match {
+        case Success(res) if cond(res) =>
+          p.trySuccess(Some(res))
+        case _ =>
+          synchronized { i.getAndIncrement() }
```

Comment:
Sorry, I wasn't very clear with my previous comment: you should remove the `synchronized` usage.

Also, you can probably use `incrementAndGet` to ensure `trySuccess` only gets called once (although I don't know whether that's actually important):
```scala
val finished = i.incrementAndGet()
if (finished == n) p.trySuccess(None)
```

Reply:
No, my bad, it was clear!